### DISCUSSION TITLE
Team 7028 Task Reports

### DISCUSSION THREAD

**Discussion by agasser** [wpilibsuite/2025Beta#19] [Task Reports]
Task 1
No issues install WPILib.
NI displayed an error during the uninstall of last year, but trying again was successful.

**Comment by agasser** [comment]
Task 4
The change to the Java Units Library caused a lot of extra work. We used the library throughout our code in 2024, and there were over 100 references to manually update. I'm surprised this breaking change was made. Was there no way to follow a compatible upgrade path, deprecating this year and removing next year?
- It appears the `negate()` method was removed, this is a breaking change that wasn't documented.
- The mutable object example in the documentation needs the imports fixed.

**Reply by SamCarlberg** [reply]
> Was there no way to follow a compatible upgrade path, deprecating this year and removing next year?

Unfortunately not. The Java type system imposes a lot of constraints on us (this is also why `VelocityUnit` can inherit from `PerUnit`, but `Velocity` cannot inherit from `Per`)

> It appears the `negate()` method was removed, this is a breaking change that wasn't documented.

It was renamed to `unaryMinus` to be consistent with other WPILib math types like `Rotation2d`. I agree that a migration guide would be helpful to point this out

**Reply by agasser** [reply]
Thanks for the reply. Ideally, `negate()` would have been deprecated and removed next year. The deprecation comment could point users to `unaryMinus()`.

**Reply by sciencewhiz** [reply]
negate exists and is deprecated in beta 2

**Reply by sciencewhiz** [reply]
Can you try importing your 2024 code again with beta 2? The importer should update the imports and some of the simpler pieces of code, which should reduce how much you need to manually change.

**Reply by agasser** [reply]
I added a note to #51 with a couple more results.

It would be helpful if the conversion also handled the `zero` method to convert to the new style from:
```
MutableMeasure<Angle> pitch = MutableMeasure.zero(Radians);
```
to: 
```
MutAngle pitch = Radians.mutable(0);
```

**Reply by agasser** [reply]
Thanks for adding `negate()` back and deprecating it, but it's missing the Javadoc comment to indicate that it's replaced with `unaryMinus()`

**Reply by sciencewhiz** [reply]
the javadoc deprecation comment is there. https://github.com/wpilibsuite/allwpilib/blob/b6de7acbdb1cb0f9db0c270d0e584ef8f9050cec/wpiunits/src/main/java/edu/wpi/first/units/Measure.java#L148-L161

**Reply by agasser** [reply]
The IDE doesn't show the Javadoc on implementation classes, like `Current`. Apparently deprecations aren't inherited by overrides. The `@Deprecated` annotation is added to the child classes, but the JavaDoc `@deprecated` comment is not there, and it isn't inherited. https://github.com/wpilibsuite/allwpilib/blob/main/wpiunits/src/generated/main/java/edu/wpi/first/units/measure/Current.java#L47

Here's a screenshot from `Current`:
![image](https://github.com/user-attachments/assets/dde7d039-df06-4342-b700-c502ee24b9f6)

**Comment by agasser** [comment]
Task 4
Just starting to research, but where do I find info on the LED animation API? I don't see anything in [the docs](https://docs.wpilib.org/en/latest/docs/software/hardware-apis/misc/addressable-leds.html#addressable-leds).

**Reply by PeterJohnson** [reply]
They haven’t been merged yet.  The WIP docs are here: https://github.com/wpilibsuite/frc-docs/pull/2774

**Reply by sciencewhiz** [reply]
There's draft documentation here: https://github.com/wpilibsuite/frc-docs/pull/2774 https://frc-docs--2774.org.readthedocs.build/en/2774/docs/software/hardware-apis/misc/addressable-leds.html

**Reply by agasser** [reply]
I'm impressed by the animation API, this looks super useful, and smart how all of the patterns are stateless.
I have a slightly advanced feature request. We would want a view that wraps around the physical strip.

```
Physical:  0  1  2  3   4  5  6  7   8  9  10
           ↓  ↓  ↓  ↓   ↓  ↓  ↓  ↓   ↓  ↓  ↓
          [c, d), _, _, _, _, _, _, _, (a, b]
           ↑  ↑                         ↑  ↑ 
    View:  2  3                         0  1   
```

**Reply by KangarooKoala** [reply]
You should be able to use `LEDPattern.offsetBy(9)` or `LEDPattern.offsetBy(-2)` (which is in the Offset section of the readthedocs above):
```
Physical:  0  1  2  3  4  5  6  7  8    9  10
           ↓  ↓  ↓  ↓  ↓  ↓  ↓  ↓  ↓    ↓  ↓
          [c, d, e, f, g, h, i, j, k), (a, b]
           ↑  ↑  ↑  ↑  ↑  ↑  ↑  ↑  ↑    ↑  ↑ 
    View:  2  3  4  5  6  7  8  9  10   0  1
```
Depending on what exactly you want to do, you might be able to then use a mask to restrict the view.

**Reply by SamCarlberg** [reply]
An offset pattern will do what you want, but I'll note that it's broken when combined with effects like masks and blends until https://github.com/wpilibsuite/allwpilib/pull/6948 is merged

**Reply by agasser** [reply]
Smart, you might be onto something. The real problem is that there are two logical views, and one wraps.
The use case is a single LED strip that starts on a corner and runs around the perimeter of the robot. We want to run mirrored animations, so one view wraps around the strip:
```
Physical:  0  1  2  3  4  5  6  7  8  9
           ↑  ↑  ↑  ↑  ↑  ↑  ↑  ↑  ↑  ↑
   View0:  2  1  0                 4  3
   View1:           0  1  2  3  4
```
I think I could create 3 views, and then apply the same pattern with different offsets.

**Reply by agasser** [reply]
I have another case, but it's probably out of scope because it doesn't fit the design. It probably needs to be stateful.
We have a short boot animation that runs when the robot starts. It seems like pattern isn't designed for this sort of use because the animation has a defined start state, and runs to completion, then ends.

**Reply by SamCarlberg** [reply]
> The use case is a single LED strip that starts on a corner and runs around the perimeter of the robot. We want to run mirrored animations, so one view wraps around the strip

Hmm. I think we'll need to add a way to create a view that can wrap around a strip like that. We currently prohibit views from being created with negative offsets, but something along those lines might be what we need; either that, or a way of explicitly opting out of the automatic reversal if the starting index is after the ending index.


> We have a short boot animation that runs when the robot starts. It seems like pattern isn't designed for this sort of use because the animation has a defined start state, and runs to completion, then ends

Yes, this is something that would need to be custom code. Wrapping in a command would probably be easiest, since that already manages initialization and execution lifecycles

**Reply by agasser** [reply]
No worries, they're both odd cases.

**Reply by SamCarlberg** [reply]
I can see wrapped LED strips being a common enough usecase that it may be useful to have support for it in WPILib. Stateful animations probably not, since you can do that with a combination of commands and custom patterns; and a generic library solution likely wouldn't be broadly applicable
